waspc/src/Wasp/Generator/WriteFileDrafts.hs

```diff
@@ -0,0 +1,161 @@
+{-# LANGUAGE TupleSections #-}
+
+module Wasp.Generator.WriteFileDrafts
+  ( writeFileDrafts,
+    fileDraftsToWriteAndFilesToDelete, -- Exported for testing.
+    removeFromChecksumFile,
+  )
+where
+
+import Control.Arrow (Arrow (first))
+import Control.Monad (unless)
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Encode.Pretty as AesonPretty
+import qualified Data.ByteString.Lazy as BSL
+import Data.Either (lefts, rights)
+import Data.List (find)
+import Data.Maybe (isJust, isNothing)
+import StrongPath (Abs, Dir, File', Path', Rel, relfile, (</>))
+import qualified StrongPath as SP
+import System.Directory (removeDirectoryRecursive, removeFile)
+import System.IO.Error (isDoesNotExistError)
+import UnliftIO.Exception (catch, throwIO)
+import Wasp.Generator.Common (ProjectRootDir)
+import Wasp.Generator.FileDraft (FileDraft, write)
+import Wasp.Generator.FileDraft.Writeable (FileOrDirPathRelativeTo, Writeable (getChecksum, getDstPath))
+import Wasp.Util (Checksum)
+
+-- | Writes given file drafts to disk, in the provided destination directory.
+-- Also makes sure to remove any redundant file drafts that have been left on the disk from before.
+-- It is smart when writing, so it doesn't write file drafts that are already written on the disk from before.
```

Comment:
maybe we should reflect this smartness in the name of the function as well? Otherwise it sounds a bit unexpected that `writeFileDrafts` is not writing all the file drafts. Maybe it could be `writeFileDraftsSmart`, or `writeFileDraftsUseChecksums`, or `ensureCorrectFilesAreWrittenToDisk`, just throwing the ideas here :D

Reply:
Yeah @shayneczyzewski and I discussed it but had hard time figuring out a name!
It writes files that need writing, and removes files that are redundant.
So how to capture that in a name? Tricky!
Hmmm maybe `writeGeneratedProjectToDisk`?
Or `ensureProjectIsGeneratedOnTheDisk`?
Not sure hm!

Reply:
I'm not the best at name golf admittedly, and this is hard since it both _conditionally_ writes _and_ deletes... 🤔 what we are wanting to convey is the in-memory representation gets materialized and matches the on-disk representation- that they are in sync. So maybe `ensureGeneratedProjectIsConsistent`, or `synchronizeFileDraftsWithDisk`? Ah I dunno on this one sorry :/

Reply:
Ooooo `synchronizeFileDraftsWithDisk` is nice :D!

Reply:
haha thanks! @matijaSos does `synchronizeFileDraftsWithDisk` sound better to you?